### PR TITLE
Only send group membership changes to Okta

### PR DESCRIPTION
**Pull request by bmaden-usds**
## Related Issue or Background Info

- https://github.com/CDCgov/prime-simplereport/issues/1296

## Changes Proposed

- Alter `updateUserPrivileges` so it only calls Okta for facilities that have been added/removed from the user, using a diff of the user’s old and new facilities

## Additional Information

- There is still a fair amount of okta usage, but now only the requested group membership changes are done instead of removing all org groups and then re-adding
- Other behavior is (mostly) retained
  - Remove user from all org groups if they aren't in the default group for some reason
  - A difference is that if we try to add a user to a non-existent group, they will now remain in the org groups that overlap (were in both the user's existing groups and the newly requested groups).  I think this is okay because it should only occur for a misconfigured organization.

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 1]
```diff
@@ -222,56 +222,72 @@ public LiveOktaRepository(
...
-          g.removeUser(user.getId());
-        }
-      }
-    }
-    if (!userInOrg) {
+
+    final String groupOrgPrefix = generateGroupOrgPrefix(orgId);
+    final String groupOrgDefaultName = generateRoleGroupName(orgId, OrganizationRole.getDefault());
+
+    Map<String, Group> currentOrgGroupMapForUser =
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:229]: nit: could you add a comment to explain what this String represents? At first glance, it's hard to find.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 2]
```diff
@@ -222,56 +222,72 @@ public LiveOktaRepository(
...
+        user.listGroups().stream()
+            .filter(
+                g ->
+                    GroupType.OKTA_GROUP.equals(g.getType())
+                        && g.getProfile().getName().startsWith(groupOrgPrefix))
+            .collect(Collectors.toMap(g -> g.getProfile().getName(), g -> g));
+
+    if (!currentOrgGroupMapForUser.containsKey(groupOrgDefaultName)) {
+      currentOrgGroupMapForUser.values().forEach(g -> g.removeUser(user.getId()));
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:238]: It seems like this is clearing `currentOrgGroupMapForUser`, but why is that necessary since we're throwing an exception?

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 3]
```diff
@@ -222,56 +222,72 @@ public LiveOktaRepository(
...
-          String.format("Cannot add Okta user to nonexistent organization=%s", orgId));
+    // to remove...
+    Set<String> groupNamesToRemove = new HashSet<>(currentOrgGroupMapForUser.keySet());
+    groupNamesToRemove.removeIf(expectedOrgGroupNamesForUser::contains);
+
+    // to add...
+    Set<String> groupNamesToAdd = new HashSet<>(expectedOrgGroupNamesForUser);
+    groupNamesToAdd.removeIf(currentOrgGroupMapForUser::containsKey);
+
+    if (groupNamesToRemove.size() + groupNamesToAdd.size() > 0) {
```
Comment (emmastephenson) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:262]: also a nit, but consider using `isEmpty` over size > 0.

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 2 above)
Comment (brmaden) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:238]: This is to retain the behavior the method previously had.  It covers the case where the user had some organizational group memberships, but was not a member of their `NO_ACCESS` group.  Since that is an invalid configuration, this will purge their remaining org-level group memberships.

Good point, it isn't obvious why this is done.

**Review by emmastephenson — APPROVED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 4]
```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
...
+    final String groupOrgPrefix = generateGroupOrgPrefix(orgId);
+    final String groupOrgDefaultName = generateRoleGroupName(orgId, OrganizationRole.getDefault());
+
+    // Map user's current Okta group memberships (Okta group name -> Okta Group).
+    // The Okta group name is our friendly role and facility group names
+    Map<String, Group> currentOrgGroupMapForUser =
+        user.listGroups().stream()
+            .filter(
+                g ->
+                    GroupType.OKTA_GROUP.equals(g.getType())
```
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:235]: optional nit: because these are enums, == works fine

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 5]
```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
...
+    groupNamesToRemove.removeIf(expectedOrgGroupNamesForUser::contains);
+
+    // to add...
+    Set<String> groupNamesToAdd = new HashSet<>(expectedOrgGroupNamesForUser);
+    groupNamesToAdd.removeIf(currentOrgGroupMapForUser::containsKey);
+
+    if (!groupNamesToRemove.isEmpty() || !groupNamesToAdd.isEmpty()) {
+      Map<String, Group> fullOrgGroupMap =
+          _client.listGroups(groupOrgPrefix, null, null).stream()
+              .filter(g -> GroupType.OKTA_GROUP.equals(g.getType()))
```
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:269]: ditto re ==

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 6]
```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
...
+              .collect(Collectors.toMap(g -> g.getProfile().getName(), g -> g));
```
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:270]: if you wanted, you could replace g -> g with Function.identity()

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 7]
```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
...
+      if (fullOrgGroupMap.size() == 0) {
+        throw new IllegalGraphqlArgumentException(
+            String.format("Cannot add Okta user to nonexistent organization=%s", orgId));
+      }
+
+      for (String groupName : groupNamesToRemove) {
+        Group group = fullOrgGroupMap.get(groupName);
+        LOG.info("Removing {} from Okta group: {}", username, group.getProfile().getName());
+        user.addToGroup(group.getId());
```
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:279]: this should be _*removing*_ the user from the group, not adding, correct?

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 8]
```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
...
+
+      groupNamesToAdd.stream()
+          .filter(n -> !fullOrgGroupMap.containsKey(n))
+          .forEach(
+              n -> {
+                throw new IllegalGraphqlArgumentException(
+                    String.format("Cannot add Okta user to nonexistent group=%s", n));
+              });
+      for (String groupName : groupNamesToAdd) {
+        Group group = fullOrgGroupMap.get(groupName);
```
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:290]: nit: would be a touch easier to read if you just added `if !fullOrgGroupMap.containsKey(groupName) throw new...` after line 289, and removed the stream right above

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java [hunk 9]
```diff
@@ -222,56 +222,76 @@ public LiveOktaRepository(
...
+        LOG.info("Adding {} to Okta group: {}", username, group.getProfile().getName());
+        group.removeUser(user.getId());
```
Comment (jeremyzitomer-usds) [inline backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:292]: this should be _adding_ the user, not removing, right?

**Review by jeremyzitomer-usds — CHANGES_REQUESTED** [review]

File: backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java (hunk 7 above)
Comment (bmaden-usds) [reply backend/src/main/java/gov/cdc/usds/simplereport/idp/repository/LiveOktaRepository.java:279]: You are absolutely correct.

**Review by jeremyzitomer-usds — APPROVED** [review]
